uportal-war/src/main/java/org/jasig/portal/events/LoginEvent.java

```diff
@@ -64,6 +67,40 @@ private LoginEvent() {
         }
         this.attributes = attributesBuilder.build();
     }
+
+    /**
+     * Instantiate a new LoginEvent from its raw elements.
+     *
+     * @param groups non-null Set of String group names of which the user who logged in is member
+     * @param userAttributes non-null user attributes of the user who logged in
+     * @param source non-null source of the LoginEvent
+     * @param serverName non-null name of the server logged in to
+     * @param eventSessionId on-null id of the session the user logged in to
+     * @param person non-null person who has logged in
+     * @param portalRequest non-null request in which this login has happened
+     *
+     * @since uPortal 4.2
+     */
+    public LoginEvent (
```

Comment:
Yeah, `LoginEvent` was weirdly obnoxious in that it's a public class and in concept new things should be able to instantiate `LoginEvents` and fire them, and yet it was hard to instantiate.  The existing package-private constructor is effectively `protected`, because it takes as an argument a which is a `protected static class` in the super-class (`PortalEvent`).

I might be missing some kind of intended elegance here, but adding a public constructor that simply accepts the state that goes into a `LoginEvent` as constructor arguments, well, that seems a lot more reasonable, and makes it handy to instantiate these for unit testing.